ateam_ai/src/plays/defense.hpp

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 A Team
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#ifndef PLAYS__DEFENSE_HPP_
+#define PLAYS__DEFENSE_HPP_
+
+#include <Eigen/Dense>
+#include <array>
+#include <vector>
+#include <optional>
+
+#include "types/world.hpp"
+#include "types/behavior_goal.hpp"
+#include "ateam_geometry/utilities.hpp"
+#include "ateam_geometry/ateam_geometry.hpp"
+
+const FieldSidedInfo & our_side_info
+DirectedGraph<BehaviorGoal> generate_basic_defense(
+  const World & world,
+  const Feild & field,
+  const FieldSidedInfo & our_side_info)
+{
+  DirectedGraph<BehaviorGoal> defense_graph;
+  // Go to the middle of the goalie area
+  BehaviorGoal goalie = get_goalie_behavior_goal(our_side_info);
+  defense_graph.add_node(goalie);
+
+  std::vector<BehaviorGoal> defenders = get_defense_behavior_goals(world, field, 2);
+  for (BehaviorGoal defender : defenders) {
+    defense_graph.add_node(defender);
+  }
+  return defense_graph;
+}
+
+BehaviorGoal get_goalie_behavior_goal(const FieldSidedInfo & our_side_info)
+{
+  // Line that is 0.5 m from the defense area in all directions
+  ateam_geometry::Segment goalie_line = ateam_geometry::Segment(
+    ateam_geometry::Point(
+      -4,
+      0.5), ateam_geometry::Point(
+      -4, -0.5));
```

Comment:
We'll have to confirm the field size is correct for div b if we haven't already. Grsim may default to div a

Reply:
As long as our coordinates are in meters, I think this is correct:
https://robocup-ssl.github.io/ssl-rules/sslrules.html#_dimensions
(The defense area is [3.5, 4.5] m in the x direction and [-1, 1] in the y direction on each team's side on the Div B field)

Reply:
> As long as our coordinates are in meters, I think this is correct: https://robocup-ssl.github.io/ssl-rules/sslrules.html#_dimensions (The defense area is [3.5, 4.5] m in the x direction and [-1, 1] in the y direction on each team's side on the Div B field)

So for this function, we simply move 0.5 m from each of those boundaries :)

Reply:
Ezpz